### PR TITLE
Adds luminol.

### PR DESCRIPTION
**Pull request by DrSnips**
[content]
Requested by @otatoh 
This is a part of an update to detectives. Luminol will glow a bright blue color when applied to items that have had blood on them at some point even if it was washed off. How useful this information can be is up to the user. The best way to apply it is by spraying it and I plan to add a spray bottle of it in the before mentioned detective update. Luminol itself can be washed away like any other form of blood or mess.
It can be made with 3 parts space cleaner and 8 parts carbon resulting in 3 parts luminol, 1 part water, and 6 parts hydrogen. It has this silly recipe due the the conflicts with other recipes like space cleaner.
**I've likely missed a few interactions and I encourage people to report them as they pop up. This also has not been live tested with perma-gunk (due to not having any way of testing it) so I have no idea how that will go.**
luminol

:cl:
 * rscadd: Added Luminol. You can spray, apply, etc. it to stuff to make it glow blue to indicate if something had blood on it. i.e. floors and items. It can be made with 3 parts space cleaner and 8 parts carbon resulting in 10 parts luminol.
* experiment: Some interactions with luminol might be unaccounted for. Please report any issues in it's behavior that seem incorrect.

### CODE REVIEW

**Comment by Kurfursten** [comment]
Would you go into more detail with why you chose such a strange recipe?

**Comment by DeityLink** [comment]
You forgot runes. It'd be interesting if the floor remembered every cult word ever written on it, so it'd show them when sprayed with luminol instead of just splatters.

**Comment by ghost** [comment]
That is a horrible recipe.

How you coded it currently means that it will always function even if diluted to tiny amounts so it's ultimately irrelevant how bad the recipe is and will only require a burden of knowledge to use efficiently. Though this is not obvious from any descriptions and will require people to read code.

Make it be space 3 cleaner + 8 carbon = 11 luminol. Nobody likes side products in recipes.

**Comment by DrSnips** [comment]
The strange recipe was due to the overlap with space cleaner if you added just the elements used in luminol (C8H7N3O2) it reacts into space cleaner due to ammonia in water reacting before it. I can change the resulting amount I guess. I just thought I would get more flak for it being even more incorrect if I made the result more than chemically possible.

**Comment by DrSnips** [comment]
As for runes I have not tested if they interact and since I don't think adding more data to every tile to hold onto forever is a good idea I think it's better to not.

**Comment by ghost** [comment]
It doesn't need to be chemically accurate, this is a game with daily dose chem and explosive bananas.

**Comment by DrSnips** [comment]
I personally don't think the recipe is that strange or hard to make so I'm not going to change that but I will increase the product to 10 and remove the current side products.

**Comment by Sooose** [comment]
Unfortunately I don't think anyone cares about blood being anywhere after gunk.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 1]
```diff
@@ -353,6 +353,14 @@ emp_act
 		w_uniform.add_blood(source)
 		update_inv_w_uniform(update)
 
+/mob/living/carbon/human/apply_luminol(var/update = FALSE) //Dispite what you might think with FALSE this will update things as normal.
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Despite

Comment (jknpj) [reply code/modules/mob/living/carbon/human/human_defense.dm]: Literally unplayable

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 2]
```diff
@@ -7152,12 +7152,9 @@ var/global/list/tonio_doesnt_remove=list("tonio", "blood")
 /datum/reagent/luminol
 	name = "Luminol"
 	id = LUMINOL
-	description = "Luminol."
+	description = "A chmical that exhibits chemiluminescence in the presence of blood due to the iron and copper in the hemoglobin."
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/Chemistry-Reagents.dm]: Chmical (10 units)

Comment (DrSnips) [reply code/modules/reagents/Chemistry-Reagents.dm]: I injected many marijuanas while coding.

**Comment by brndd** [comment]
Update the changelog to match the new recipe.

File: code/game/atoms.dm [hunk 3]
```diff
@@ -750,6 +751,18 @@ its easier to just keep the beam vertical.
 	if(fingerprintshidden && istype(fingerprintshidden))
 		A.fingerprintshidden |= fingerprintshidden.Copy()    //admin	A.fingerprintslast = fingerprintslast
 
+//Atomic level procs to be used elsewhere.
+/atom/proc/apply_luminol(var/atom/A)
```
Comment (ShiftyRail) [inline code/game/atoms.dm:755]: Can replace the whole proc with `return has_blood`

File: code/game/atoms.dm [hunk 4]
```diff
@@ -750,6 +751,18 @@ its easier to just keep the beam vertical.
...
+	if(had_blood)
+		return TRUE
+	else
+		return FALSE
+
+/atom/proc/clear_luminol(var/atom/A)
```
Comment (ShiftyRail) [inline code/game/atoms.dm:758]: Same as above

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 5]
```diff
@@ -42,11 +42,33 @@
...
 	gender = PLURAL
 	density = 0
-	anchored = 1
-	luminosity = 1
+	anchored = TRUE
 	icon = 'icons/effects/effects.dmi'
 	icon_state = "greenglow"
 
+/obj/effect/decal/cleanable/greenglow/New()
+	..()
```
Comment (ShiftyRail) [inline code/game/objects/effects/decals/Cleanable/misc.dm:50]: Prefer `. = ..()` as it is slightly less likely to fuck up in the future but not strictly needed.

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 6]
```diff
@@ -42,11 +42,33 @@
...
+	density = 0
+	anchored = TRUE
+	icon = 'icons/effects/blood.dmi'
+	icon_state = "mfloor1"
+	//icon = 'icons/effects/tomatodecal.dmi'
+	//icon_state = "fruit_smudge1"
+	color = LIGHT_COLOR_CYAN
+
+/obj/effect/decal/cleanable/blueglow/New()
+	..()
```
Comment (ShiftyRail) [inline code/game/objects/effects/decals/Cleanable/misc.dm:66]: Same as above.

File: code/game/objects/items.dm [hunk 7]
```diff
@@ -1054,6 +1055,34 @@ var/global/list/image/blood_overlays = list()
 
 	blood_overlays[type] = image(I)
 
+/obj/item/apply_luminol()
+	if(!..())
+		return FALSE
+	if(!blood_overlays[type]) //Blood overlay generation if it lacks one.
+		generate_blood_overlay()
+	if(blood_overlay != null)
```
Comment (ShiftyRail) [inline code/game/objects/items.dm]: `if(blood_overlay)` is more general. This will fail if `blood_overlay` is zero or an empty string.

**Review by ShiftyRail — DISMISSED** [review]

File: code/game/objects/effects/decals/Cleanable/misc.dm (hunk 5 above)
Comment (DrSnips) [reply code/game/objects/effects/decals/Cleanable/misc.dm:50]: I can't say I've seen it cause issues before nor have I heard of it causing issues. I'm open to changing it if you can give me a bit more than personal taste.

File: code/game/atoms.dm (hunk 3 above)
Comment (DrSnips) [reply code/game/atoms.dm:755]: Resolved.

File: code/game/atoms.dm (hunk 4 above)
Comment (DrSnips) [reply code/game/atoms.dm:758]: Resolved.

File: code/game/objects/items.dm (hunk 7 above)
Comment (DrSnips) [reply code/game/objects/items.dm]: Resolved.

File: code/modules/reagents/Chemistry-Recipes.dm [hunk 8]
```diff
@@ -989,6 +989,17 @@
 	required_reagents = list(SODIUMCHLORIDE = 2, CLEANER = 2, OXYGEN = 1)
 	result_amount = 1
 
+//This one isn't even close the the real life reaction but will have to do to avoid conflicts with the above reactions.
+/datum/chemical_reaction/luminol
+	name = "Luminol"
+	id = LUMINOL
+	result = LUMINOL
+	required_reagents = list(CLEANER = 3, CARBON = 8)
+	result_amount = 10
```
Comment (brndd) [inline code/modules/reagents/Chemistry-Recipes.dm:998]: Your changelog doesn't match this, it still has the "realistic" results with byproduct water and hydrogen.

**Review by brndd — CHANGES_REQUESTED** [review]

File: code/modules/reagents/Chemistry-Recipes.dm (hunk 8 above)
Comment (DrSnips) [reply code/modules/reagents/Chemistry-Recipes.dm:998]: OP updated.

File: code/game/objects/effects/decals/Cleanable/misc.dm (hunk 5 above)
Comment (Kurfursten) [reply code/game/objects/effects/decals/Cleanable/misc.dm:50]: what . = ..() means it that it sets the return value to the return value of the parent, it's useful if you want to for example do something like

```
. = ..()
CODE SPECIFIC TO THIS CHILD WHICH NEEDS TO OPERATE AFTER THE PARENT'S CODE IS EXECUTED
return .
```
Allows you to return ..() without immediately ending the proc. In short, it's not necessary if you aren't returning anything like in a New()

**Comment by Kurfursten** [comment]
It is actually totally possible to test permagunk on local but I'm gonna let this slide. I am expecting you to keep an eye on this feature and bug reports though.

**Pull request merged**